rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,223 @@
+    can be easily defined by a subgraph developer ultimately providing 
+    the tools necesssary for front end developers to create effective 
+    user interfaces without the need to middleware.
+     
+## Urgency
+
+Fulltext search filters are not essential to the value proposition of 
+the Graph Node, so the implementation is not urgent. However, several 
+high profile users have communicated that it may be a conversion blocker, 
+so it would be a valuable addition and should be added immediately. 
+
+## Terminology
+
+- `lexeme`: a basic lexical unit of a language, consisting of one word or 
+  several words, considered as an abstract unit, and applied to a family 
+  of words related by form or meaning.
+- `morphology (linguistics)`: the study of words, how they are formed, 
+  and their relationship to other words in the same language. 
+- `fulltext search index`: the result of lexical and morphological 
+  analysis (stemming) of a set of text documents.  It provides frequency 
+  and location for the language-specific stems found in the text documents 
+  being indexed. 
+
+## Detailed Design
+
+### Subgraph Definition
+
+Part of the power of fulltext search indexes are their flexibility, so 
+it is important to expose a simple interface to facilitate useful 
+applications of the index and ulimately aim to reduce the need to create 
+new subgraphs for the express purpose of updating fulltext search fields 
+for end-use cases. 
+
+For each fulltext search index a subgraph developer must be able to specify:
+    1. a language, 
+    2. a set of text document fields to include,
+    3. relative weighting for each field in the index.
+
+The proposed process of adding a fulltext search index to an Entity 
+involves adding a special field to that Entity in the GraphQL schema of 
+the subgraph's definition. The subgraph developer will create a new virtual 
+field of type String that is marked with the special `@FullTextSearch` directive. 
+The directive will have two required parameters for specifying the language
+ and the fields to be incorporated. Verification of the index definition 
+ will ensure that all fields referenced are valid String type fields on 
+ that entity. To be clear, the proposed interface will add the fulltext 
+ search definitions to the subgraph deployment hash, so adding or updating 
+ a fulltext search index will require an update to the subgraph.  
+ With subgraph composition it will be possible to easily create new subgraphs 
+ that add specific fulltext search indexes to an existing subgraph. 
+
+Example fulltext search field definition:
+```graphql
+type ExampleItem @entity {
+    id: ID!,
+    name: String!,
+    description: String!,
+    specs: String!,
+    address: Bytes!,
+    marked: Boolean!,
+    itemSearch: String! @fullTextSearchIndex(language: "english", fields: [{name: "name", weight: 5}, {name: "description", weight: 3}, {name: "specs", weight: 1}])
```

Comment:
I thought we decided to move away from including this in the schema. When did we decide to put it back?

Reply:
When you say "move away from including this in the schema" I'm assuming you are also talking about the decision to include it in the subgraph definition in the schema? Or are you simply saying it should be in the subgraph definition, but somewhere else?

Reply:
I have definitely gone back and forth on putting it in the base subgraph definition or allowing it to be an add on to an already deployed subgraph, but we had a brief discussion a few weeks ago on a call; Jannis and I (if I remember correctly) agreed that it needs to be a part of the definition of uniqueness for a subgraph to keep a stable GraphQL API for each subgraph. It won't work to have a query filter or type change out from under a user if indexes are updated. With the subgraph composition though we will be able to maintain flexibility for subgraph developers to add indexes to existing subgraphs.  

Looking forward to the decentralized network, the more I thought through the added complexity of essentially adding an indexing marketplace to the query marketplace the more it seems that it poses risks to the health of the data network. 
  - Will index updates have their own system for being propagated to Graph Nodes storing that subgraph? Why not integrate them into the definition of a subgraph, so we can leverage the existing query discovery and subgraph data propagation systems.
  -  A query engine will have to find a node that not only has the subgraph deployment, but also the correct combination of indexes applied to it.  This will involve propagating index updates to the network and could result in a less clear graph of data.  
- On the other hand you could argue that forcing users to create a new subgraph in order to update or add a fulltext search index will lead to a large set of very similar subgraphs, but I _think_ the complexity of organizing that through interface design will in the end be simpler than managing different schemas/indexes on a single subgraph.  I also think the aforementioned interface work will be useful or needed for managing the complexity introduced by subgraph composition which could also leading to situations where many similar subgraphs are being created.
  - Perhaps less important than correctness on the decentralized network, but keep in mind there will need to be new interfaces designed and built on Graph Explorer for managing indexes.

Concerning the choice of putting the definition in the schema: I've also thought about creating yaml files for the index definition or adding to the subgraph definition, but the schema is where the data model and query model is defined and it makes sense to keep it query concerns there.  Though it might make sense to separate raw data model concerns from front end query concerns, we are already using @derivedFrom for virtual fields that are derived from Entity data and I think keeping the query model in one place together with the data model is a useful simplification.

Reply:
Alright I see the case for this. Then can we look at these changes:
1. Switch to @index directive with a type: "fulltext"
2. Make sure the fields attributes can include fields on other entities in the schema (including imported types from subgraph composition)?

Reply:
It seems like a lot of thought has gone into this and the concerns above around discovery and alignment around the right text full-text search indexes seem valid. These are definitely different from normal attribute indexes in that they are much harder to organically align on in a distributed fashion based on query access patterns (because you still need to choose what type of search index).

Even so, combining the data model for an entity w/ a specific way of searching that entity feels wrong to me... could we still include the same information in the subgraph, but somewhere else in the manifest, perhaps as a "hint" to the Indexer rather than a mandate?

Indexes provided as hints could be given a label or an ID, which end-users can then use to reference in their query to force Indexers to use a specific index in resolving a query. This concept could be extended beyond full-text search indexes as well.

Reply:
We could define custom indexes in the subgraph manifest (yaml) rather than in the schema, but I think that makes it more difficult to manage. 
- A new index also means a new query with specific behavior, so I'm having trouble following how it would work as just "a hint to the Indexer".  
- We currently have many indexes per subgraph; they are derived from the schema.  Why change the paradigm?
- All of our query types are currently generated from information in the schema.  Why split query definitions across files in the subgraph?
- The index definition will consist mostly of entity field names to be included these are in the schema, so working within one file (while not a huge difference) will be easier. 

Rather than give indexes a unique id (which will not be self-describing) I think we should creates a unique query for the index, so end-users can use that specific query to force the Indexer to use a specific index. Indexers can implement that index using their own custom systems, but it should always search across the fields described in the schema (or manifest based on your suggestion). The query name can be defined in the manifest by the subgraph developer.